src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -5864,35 +5950,1058 @@ void Main(string? s)
+            c.VerifyTypes();
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void VarLocalFromGenericMethod_WithNullableReturn()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string? key)
+    {
+        var s = Copy(key);
+        s/*T:string?*/.ToString(); // warn
+    }
+    public T? Copy<T>(T key) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyVarLocal(c, "string?");
+            c.VerifyTypes();
+            c.VerifyDiagnostics(
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         s/*T:string?*/.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(7, 9)
+                );
+        }
+
+        [Fact]
+        public void VarLocalFromGenericMethod_WithNullableReturn_WithNonNullArgument()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string? key)
+    {
+        key = ""hello"";
+        var s = Copy(key);
+        s/*T:string?*/.ToString(); // warn
+    }
+    public T? Copy<T>(T key) => throw null;
+}
+", parseOptions: TestOptions.Regular8);
+
+            VerifyVarLocal(c, "string?");
+            c.VerifyTypes();
+            c.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         s/*T:string?*/.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void GenericMethod_WithEnsuresNotNullOnReturnType()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    [EnsuresNotNull]
+    public static T Copy<T>(T key) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (5,6): error CS0592: Attribute 'EnsuresNotNull' is not valid on this declaration type. It is only valid on 'parameter' declarations.
+                //     [EnsuresNotNull]
+                Diagnostic(ErrorCode.ERR_AttributeOnBadSymbolType, "EnsuresNotNull").WithArguments("EnsuresNotNull", "parameter").WithLocation(5, 6)
+                );
+            // PROTOTYPE(NullableReferenceTypes): Need to confirm if this would be useful
+        }
+
+        [Fact]
+        public void SuppressedNullGivesNonNullResult()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+public class C
+{
+    public void Main(string s)
+    {
+        s = null!; // null! returns an oblivious result
+        var s2 = s;
+        s2/*T:string*/.ToString(); // ok
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            // PROTOTYPE(NullableReferenceTypes): I'd expect null! to return a non-null result
```

Comment:
@cston Merging the latest bits from the feature branch caused a change here. It used to be that the type verification would expect `s2/*:string!*/`.
What is the desired behavior? Should the suppression operator result in an oblivious or a non-null nullability state?